### ISSUE TITLE
[BUG] HTML view of transformations are not showing as a flow and instead shown as class definition directly

### ISSUE THREAD

**Issue by yarnabrina** [sktime/sktime#7152]
## Code

Reused from https://github.com/sktime/sktime/discussions/5903 (I still have that question, need to resolve that sometime)

```python
from sklearn.base import BaseEstimator
from sklearn.preprocessing import RobustScaler, QuantileTransformer
from sktime.forecasting.arima import ARIMA
from sktime.forecasting.base import BaseForecaster, ForecastingHorizon
from sktime.forecasting.compose import (
    ForecastingPipeline,
    ForecastX,
    TransformedTargetForecaster,
)
from sktime.forecasting.exp_smoothing import ExponentialSmoothing
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.base import BaseTransformer
from sktime.transformations.compose import FeatureUnion, TransformerPipeline
from sktime.transformations.series.boxcox import BoxCoxTransformer, LogTransformer
from sktime.transformations.series.detrend import Detrender, Deseasonalizer
from sktime.transformations.series.difference import Differencer
from sktime.transformations.series.subset import ColumnSelect


def create_pipeline(
    endogenous_forecaster: BaseForecaster,
    prediction_horizon: ForecastingHorizon,
    future_known_exogenous: list[str] | None = None,
    future_unknown_exogenous: list[str] | None = None,
    future_unknown_forecaster: BaseForecaster | None = None,
    endogenous_transformations: list[BaseTransformer | BaseEstimator] | None = None,
    future_known_transformations: list[BaseTransformer | BaseEstimator] | None = None,
    future_unknown_transformations: list[BaseTransformer | BaseEstimator] | None = None,
) -> BaseForecaster:
    if not endogenous_transformations:
        y_pipe = endogenous_forecaster
    else:
        for counter, endogenous_transformation in enumerate(endogenous_transformations):
            if isinstance(endogenous_transformation, BaseTransformer):
                continue

            endogenous_transformations[counter] = TabularToSeriesAdaptor(endogenous_transformation)

        y_pipe = TransformedTargetForecaster(
            [TransformerPipeline(endogenous_transformations), endogenous_forecaster]
        )

    if not future_known_exogenous:
        known_X_pipe = None
    elif not future_known_transformations:
        known_X_pipe = future_known_transformations
    else:
        for counter, future_known_transformation in enumerate(future_known_transformations):
            if isinstance(future_known_transformation, BaseTransformer):
                continue

            future_known_transformations[counter] = TabularToSeriesAdaptor(
                endogenous_transformation
            )

        known_X_pipe = TransformerPipeline(
            [ColumnSelect(columns=future_known_exogenous), *future_known_transformations]
        )

    if not future_unknown_exogenous:
        unknown_X_trafo_pipe = None
    elif not future_unknown_transformations:
        unknown_X_trafo_pipe = None
    else:
        for counter, future_unknown_transformation in enumerate(future_unknown_transformations):
            if isinstance(future_unknown_transformation, BaseTransformer):
                continue

            future_unknown_transformations[counter] = TabularToSeriesAdaptor(
                endogenous_transformation
            )

        unknown_X_trafo_pipe = TransformerPipeline(
            [ColumnSelect(columns=future_unknown_exogenous), *future_unknown_transformations]
        )

    if not future_unknown_exogenous:
        unknown_X_fcst_pipe = None
    elif not future_unknown_forecaster:
        unknown_X_fcst_pipe = endogenous_forecaster
    else:
        unknown_X_fcst_pipe = future_unknown_forecaster

    e2e_trafo_pipe_steps: list[BaseTransformer] = []

    if known_X_pipe and not unknown_X_trafo_pipe:
        e2e_trafo_pipe_steps.append(known_X_pipe)
    elif not known_X_pipe and unknown_X_trafo_pipe:
        e2e_trafo_pipe_steps.append(unknown_X_trafo_pipe)
    elif known_X_pipe and unknown_X_trafo_pipe:
        e2e_trafo_pipe_steps.append(FeatureUnion([known_X_pipe, unknown_X_trafo_pipe]))

    if e2e_trafo_pipe_steps:
        e2e_trafo_pipe = TransformerPipeline(e2e_trafo_pipe_steps)
    else:
        e2e_trafo_pipe = None

    if not unknown_X_fcst_pipe:
        e2e_fcst_pipe = y_pipe
    else:
        e2e_fcst_pipe = ForecastX(
            y_pipe,
            unknown_X_fcst_pipe,
            fh_X=prediction_horizon,
            forecaster_X_exogeneous="complement",
        )

    if e2e_trafo_pipe:
        e2e_pipe = ForecastingPipeline([e2e_trafo_pipe, e2e_fcst_pipe])
    else:
        e2e_pipe = e2e_fcst_pipe

    return e2e_pipe


create_pipeline(
    ARIMA(),
    ForecastingHorizon(values=range(1, 4)),
    future_known_exogenous=["alpha", "beta"],
    future_unknown_exogenous=["gamma", "delta"],
    future_unknown_forecaster=ExponentialSmoothing(),
    endogenous_transformations=[
        LogTransformer(),
        Detrender(model="additive"),
        Deseasonalizer(model="additive"),
    ],
    future_known_transformations=[RobustScaler(), BoxCoxTransformer()],
    future_unknown_transformations=[QuantileTransformer(), Differencer()],
)
```

Result

![image](https://github.com/user-attachments/assets/bd9ddbe5-2b68-4f4b-967f-1020c81755fa)

## Expectation

Not sure if it's a bug, but I was expecting a DAG flow for `TransformerPipeline` component.

**Comment by fkiraly** [comment]
I think the reason is that the `TransformerPipeline` does not hvae the DAG flow implemented - it should be inheriting it from the `BaseMetaEstimator`, but it looks like only the forecasting pipelines have the compatible html.

FYI @mateuszkasprowicz, is this correct?

**Comment by mateuszkasprowicz** [comment]
[`_sk_visual_block()`](https://github.com/sktime/sktime/blob/371b56128aa2a10361237b6185f46ea120d3be44/sktime/forecasting/compose/_pipeline.py#L282-L292) that returns a [`_VisualBlock`](https://github.com/sktime/sktime/blob/371b56128aa2a10361237b6185f46ea120d3be44/sktime/utils/_estimator_html_repr.py#L18) of `serial` kind (which is responsible for nice flow repr) was implemented for `sktime.forecasting.compose._pipeline._Pipeline`. `TransformerPipeline` doesn't inherit from it.

**Comment by mateuszkasprowicz** [comment]
Possibly, we can think of changing [`_get_visual_block()`](https://github.com/sktime/sktime/blob/371b56128aa2a10361237b6185f46ea120d3be44/sktime/utils/_estimator_html_repr.py#L105) to check for `_steps` property and then return a `_VisualBlock` of `serial` kind. But that would require a review of pipeline-like classes and some testing.

**Comment by fkiraly** [comment]
I see! My proposed aproach would be to move the `_get_visual_block` up to `_HeterogenousMetaEstimator` then, this is more general, a mixin for classes like pipelines that have as a key attribute a list of estimators (or, alternatively, name/estimator pairs or tuples). The name of a `_steps`-like argument sits in the attribute `_steps_attr`.

I would suggest adding another attribute, or perhaps better, a tag, which descendants can set to determine whether the blocks are horizontal/parallel, or vertical/sequential.

On a side note: long-term, we wanted to move `_steps_attr` and `_steps_fitted_attr` to tags, we do no need to do it right now as this would require refactoring together with `skbase`.

**Comment by fkiraly** [comment]
On a higher level note, should the `TransformerPipeline` connected to `ARIMA` not "auto-simplify" to one of the appropriate pipelines?

**Comment by yarnabrina** [comment]
Can you please clarify what do you mean by "auto-simplify" here? What's the appropriate pipeline?

**Comment by fkiraly** [comment]
> On a higher level note, should the `TransformerPipeline` connected to `ARIMA` not "auto-simplify" to one of the appropriate pipelines?

I mean, suppose you construct something like
```python
TransformerPipeline([a, b, c]) * my_forecaster
```

Should this not automatically result in
```python
TransformedTargetForecaster([a, b, c, my_forecaster])
```
instead of the nested pipelines?

I am aware this would already introduce elements of contextual parsing, but it still would be a simple dunder operation.

**Comment by yarnabrina** [comment]
I did not know of this auto-mofification, does it not affect the parameter hierarchy, esp. affecting the keys in getting parameter names required for tuning? Unless it doesn't affect that, I think it may be better to avoid this for user friendliness.

**Comment by fkiraly** [comment]
> I did not know of this auto-mofification, does it not affect the parameter hierarchy, esp. affecting the keys in getting parameter names required for tuning?

Good point, though I think it affects no relevant API contract - tuning etc happens only once an object is constructed, and the structure and nesting of the parameters is not changed after construction. The simplification happens only at the moment of construction.

**Comment by yarnabrina** [comment]
I'll get back soon (by this weekend) on this. This potentially affects my work usage/expectations as well quite significantly.

**Comment by mateuszkasprowicz** [comment]
> I see! My proposed aproach would be to move the `_get_visual_block` up to `_HeterogenousMetaEstimator` then, this is more general, a mixin for classes like pipelines that have as a key attribute a list of estimators (or, alternatively, name/estimator pairs or tuples). The name of a `_steps`-like argument sits in the attribute `_steps_attr`.

Moved `_sk_visual_block_` from `_Pipeline` to `_HeterogenousMetaEstimator`, and changed `self._steps` to `getattr(self, self._steps_attr)`. Does it look ok now?

![image](https://github.com/user-attachments/assets/62c53e36-52db-435a-a8f2-40271628ba0c)

**Comment by yarnabrina** [comment]
@mateuszkasprowicz this looks like two transformer pipelines are sequential, which is not the case. I'd expect them to side by side or similar.

**Comment by mateuszkasprowicz** [comment]
How about this?

![image](https://github.com/user-attachments/assets/0ff1ff2a-2bea-48f7-aab7-9c0153e1cb82)

It seems that `FeatureUnion` inherits from `_HeterogenousMetaEstimator` so with the approach mentioned above it returned a `parallel` visual block

**Comment by yarnabrina** [comment]
I am happy with parallel ones, thank you.

Probably being too picky, but it'd be good to have another box outside FeatureUnion. FeatureUnion is actually an input to TransformerPipeline (the outside one) and both that TransformerPipeline and ForecastX together form ForecastingPipeline. In current form, the parent-child relation between FeatureUnoin and outer TransformerPipeline is not clear.

**Comment by mateuszkasprowicz** [comment]
@fkiraly In the meantime, I checked all classes that inherit from `_HeterogenousMetaEstimator`, and there are a few caveats:

1. `ParamFitterPipeline` - no `_steps_attr` attribute in the class, should be a combination of `param_est` and `transformers` (?)
2.  `PwTrafoPanelPipeline` - no `_steps_attr` attribute in the class, should be a combination of `pw_trafo` and `transformers` (?)
3. `GroupbyCategoryForecaster` - there's a `_steps_attr` attribute in the class, but points to "_forecasters" which is a list of BaseForecasters instead of an iterable of (name: str, estimator, ...) tuples. Further, `transformer` argument is not a part of this list.
4.  `TransformSelectForecaster` - as above
5. `RegressorPipeline` -  no `_steps_attr` attribute in the class, should be a combination of `regressor` and `transformers` (?)
6. `ClassifierPipeline` -  no `_steps_attr` attribute in the class, should be a combination of `classifier` and `transformers` (?)
7. `ClustererPipeline` -  no `_steps_attr` attribute in the class, should be a combination of `clusterer` and `transformers` (?)

**Comment by fkiraly** [comment]
Hm, thanks for the great analysis, @mateuszkasprowicz!

Not what I would have expected.

I think there are two categories and two types of solutions here.

Category one: one estimator plus a transformer pipeline - 1, 2, 5, 6, 7

In this case, the pipeline overrides parts of the `_HeterogenousMetaestimator` and does not really use the `_steps` attribute (it is inherited, so points to `_steps`. I wonder whether the extender design was well thought out here.

What we could do, is introduce a `_steps` attribute just for the visualisation, and perhaps later other things. I would implement it as a `property`, that runs `self._check_steps` on a property `steps`. The property `steps` would be different for each estimator, I would make it the concatenation of `transformers` and then `param_est` or similar, i.e., a list of length number of transformers plus 1.

The display in category one should be sequential.

Category two: steps_attr is not in the right format - 3, 4, 8.

Here, I would do the same as in category one, but also change `_steps_attr` to point to the new attribute that has list of pairs.

The display in category two should be parallel.


Overall, I would also make separate PR for category one and two, since I'm not sure if it will be easy.

**Comment by mateuszkasprowicz** [comment]
Moved this to a separate issue: https://github.com/sktime/sktime/issues/7222 IMO to be resolved before this to have a general solution to this.

**Comment by fkiraly** [comment]
hmm, interesting! Looking forward to your suggestion!

**Comment by mateuszkasprowicz** [comment]
the idea is if we have a guarantee that every object with a `_steps_attr` should have a `serial` type of `_VisualBlock` we can update `_get_visual_block` (which is basically a default dispatcher for type of a `_VisualBlock`) with smth like this:
```python
if hasattr(base_object, "_steps_attr") and not isclass(base_object):
        base_objects = [
            (key, est)
            for key, est in getattr(base_object, "_steps_attr")
            if hasattr(est, "get_params") and hasattr(est, "fit") and not isclass(est)
        ]
        if base_objects:
            return _VisualBlock(
                "parallel",
                [est for _, est in base_objects],
                names=[f"{key}: {est.__class__.__name__}" for key, est in base_objects],
                name_details=[str(est) for _, est in base_objects],
            )
```
Will test this approach soon.

There's still a topic of `FeatureUnion` which inherits from `_HeterogenousMetaEstimator` and thus has a `_steps_attr` but should be a `parellel` type. We can always override `_sk_visual_block_ ()` so it's no big deal just pointing this out.

**Comment by fkiraly** [comment]
> the idea is if we have a guarantee that every object with a `_steps_attr` should have a `serial` type of `_VisualBlock` we can update `_get_visual_block` (which is basically a default dispatcher for type of a `_VisualBlock`) with smth like this:

We do not have this guarantee!

I would control this via a tag, i.e., using `get_tag` and a tag in the estimator, or an extra attr. The default should perhaps be parallel rather than sequential.

I would also not override `_sk_visual_block` higher than in `_HeterogenousMetaestimator`, in that case we can always assume that `_steps_attr` exists (per inheritance).
